util/src/test-support/java/tech/pegasys/artemis/util/async/AsyncRunnerTest.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.async;
+
+import static tech.pegasys.artemis.util.async.SafeFuture.propagateResult;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class AsyncRunnerTest implements AsyncRunner {
```

Comment:
I would have thought this would be `StubAsyncRunner` and it would still delay running the provided action, but would do so until the test called an extra method to execute.  So an implementation like:

```
public class StubAsyncRunner implements AsyncRunner {
  private List<Runnable> queuedActions = new ArrayList<>();

  @Override
  public <U> SafeFuture<U> runAsync(final Supplier<SafeFuture<U>> action, final Executor executor) {
    final SafeFuture<U> result = new SafeFuture<>();
    queuedActions.add(
        () -> {
          try {
            propagateResult(action.get(), result);
          } catch (final Throwable t) {
            result.completeExceptionally(t);
          }
        });
    return result;
  }

  @Override
  public <U> SafeFuture<U> runAfterDelay(
      Supplier<SafeFuture<U>> action, long delayAmount, TimeUnit delayUnit) {
    return runAsync(action, null); // Executor is ignored anyway.
  }

  public void executeQueuedActions() {
    final List<Runnable> actionsToExecute = queuedActions;
    queuedActions = new ArrayList<>();
    actionsToExecute.forEach(Runnable::run);
  }

  public boolean hasDelayedActions() {
    return !queuedActions.isEmpty();
  }
}
```

and then `Eth1DataManager.catchStartup_retryStartup` becomes:
```
  @Test
  void cacheStartup_retryStartup() {
    Request mockRequest = mockFailedRequest();
    when(web3j.ethGetBlockByNumber(any(), eq(false))).thenReturn(mockRequest);

    eth1DataManager.start();

    verify(web3j).ethGetBlockByNumber(any(), eq(false)); // First failed attempt

    for (int i = 2; i <= Constants.ETH1_CACHE_STARTUP_RETRY_GIVEUP; i++) {
      asyncRunner.executeQueuedActions();
      verify(web3j, times(i)).ethGetBlockByNumber(any(), eq(false));
    }

    assertThat(asyncRunner.hasDelayedActions()).isFalse();
  }
```

The main reason for this is that it preserves the important property that `AsyncRunner` doesn't execute the action immediately which we want to test.

Reply:
Makes sense. Done.